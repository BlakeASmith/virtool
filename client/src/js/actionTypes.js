/**
 * Created by igboyes on 03/05/17.
 */

const createRequestActionType = (root) => {
    return {
        REQUESTED: `${root}_REQUESTED`,
        SUCCEEDED: `${root}_SUCCEEDED`,
        FAILED: `${root}_FAILED`
    };
};

// Virus actionTypes
export const WS_UPDATE_VIRUS = "WS_UPDATE_VIRUS";
export const WS_REMOVE_VIRUS = "WS_REMOVE_VIRUS";

export const FIND_VIRUSES = createRequestActionType("FIND_VIRUSES");

export const GET_VIRUS_REQUESTED = "GET_VIRUS_REQUESTED";
export const GET_VIRUS_SUCCEEDED = "GET_VIRUS_SUCCEEDED";
export const GET_VIRUS_FAILED = "GET_VIRUS_FAILED";

export const CREATE_VIRUS_SET_NAME = "CREATE_VIRUS_SET_NAME";
export const CREATE_VIRUS_SET_ABBREVIATION = "CREATE_VIRUS_SET_ABBREVIATION";
export const CREATE_VIRUS_CLEAR = "CREATE_VIRUS_CLEAR";
export const CREATE_VIRUS_REQUESTED = "CREATE_VIRUS_REQUESTED";
export const CREATE_VIRUS_SUCCEEDED = "CREATE_VIRUS_SUCCEEDED";
export const CREATE_VIRUS_FAILED = "CREATE_VIRUS_FAILED";

export const EDIT_VIRUS_REQUESTED = "EDIT_VIRUS_REQUESTED";
export const EDIT_VIRUS_SUCCEEDED = "EDIT_VIRUS_SUCCEEDED";
export const EDIT_VIRUS_FAILED = "EDIT_VIRUS_FAILED";

export const REMOVE_VIRUS_REQUESTED = "REMOVE_VIRUS_REQUESTED";
export const REMOVE_VIRUS_SUCCEEDED = "REMOVE_VIRUS_SUCCEEDED";
export const REMOVE_VIRUS_FAILED = "REMOVE_VIRUS_FAILED";

// History actionTypes
export const FIND_HISTORY_REQUESTED = "FIND_HISTORY_REQUESTED";
export const FIND_HISTORY_SUCCEEDED = "FIND_HISTORY_SUCCEEDED";
export const FIND_HISTORY_FAILED = "FIND_HISTORY";

export const REMOVE_CHANGE_REQUESTED = "REMOVE_CHANGE_REQUESTED";
export const REMOVE_CHANGE_SUCCEEDED = "REMOVE_CHANGE_SUCCEEDED";
export const REMOVE_CHANGE_FAILED = "REMOVE_CHANGE_FAILED";

// Users actionTypes
export const LOAD_USERS = "LOAD_USERS";
export const CREATE_USER = "CREATE_USER";

// Account actionTypes
export const GET_ACCOUNT_REQUESTED = "GET_ACCOUNT_REQUESTED";
export const GET_ACCOUNT_SUCCEEDED = "GET_ACCOUNT_SUCCEEDED";
export const GET_ACCOUNT_FAILED = "GET_ACCOUNT_FAILED";
export const LOGOUT_REQUESTED = "LOGOUT_REQUESTED";
export const LOGOUT_SUCCEEDED = "LOGOUT_SUCCEEDED";

// Administrative setting actionTypes
export const GET_SETTINGS = createRequestActionType("GET_SETTINGS");
export const UPDATE_SETTINGS = createRequestActionType("UPDATE_SETTINGS");

export const SET_SOURCE_TYPE_VALUE = "SET_SOURCE_TYPE_VALUE";
export const SET_CONTROL_READAHEAD_TERM = "SET_CONTROL_READAHEAD_TERM";
export const GET_CONTROL_READAHEAD_REQUESTED = "GET_CONTROL_READAHEAD_REQUESTED";
export const GET_CONTROL_READAHEAD_SUCCEEDED = "GET_CONTROL_READAHEAD_SUCCEEDED";
export const GET_CONTROL_READAHEAD_FAILED = "GET_CONTROL_READAHEAD_FAILED";

export const LIST_USERS = createRequestActionType("LIST_USERS");
export const SELECT_USER = createRequestActionType("SELECT_USER");
export const CHANGE_SET_PASSWORD = "CHANGE_SET_PASSWORD";
export const CLEAR_SET_PASSWORD = "CLEAR_SET_PASSWORD";
export const SET_FORCE_RESET = createRequestActionType("SET_FORCE_RESET");
