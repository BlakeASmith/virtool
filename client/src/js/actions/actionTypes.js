/**
 * Created by igboyes on 03/05/17.
 */

// Virus actionTypes
export const WS_UPDATE_VIRUS = "WS_UPDATE_VIRUS";
export const WS_REMOVE_VIRUS = "WS_REMOVE_VIRUS";

export const FIND_VIRUSES_REQUESTED = "FIND_VIRUSES_REQUESTED";
export const FIND_VIRUSES_SUCCEEDED = "FIND_VIRUSES_SUCCEEDED";
export const FIND_VIRUSES_FAILED = "FIND_VIRUSES_FAILED";

export const GET_VIRUS_REQUESTED = "GET_VIRUS_REQUESTED";
export const GET_VIRUS_SUCCEEDED = "GET_VIRUS_SUCCEEDED";
export const GET_VIRUS_FAILED = "GET_VIRUS_FAILED";

export const CREATE_VIRUS_SET_NAME = "CREATE_VIRUS_SET_NAME";
export const CREATE_VIRUS_SET_ABBREVIATION = "CREATE_VIRUS_SET_ABBREVIATION";
export const CREATE_VIRUS_CLEAR = "CREATE_VIRUS_CLEAR";
export const CREATE_VIRUS_REQUESTED = "CREATE_VIRUS_REQUESTED";
export const CREATE_VIRUS_SUCCEEDED = "CREATE_VIRUS_SUCCEEDED";
export const CREATE_VIRUS_FAILED = "CREATE_VIRUS_FAILED";

export const EDIT_VIRUS_REQUESTED = "EDIT_VIRUS_REQUESTED";
export const EDIT_VIRUS_SUCCEEDED = "EDIT_VIRUS_SUCCEEDED";
export const EDIT_VIRUS_FAILED = "EDIT_VIRUS_FAILED";

export const REMOVE_VIRUS_REQUESTED = "REMOVE_VIRUS_REQUESTED";
export const REMOVE_VIRUS_SUCCEEDED = "REMOVE_VIRUS_SUCCEEDED";
export const REMOVE_VIRUS_FAILED = "REMOVE_VIRUS_FAILED";

// History actionTypes
export const FIND_HISTORY_REQUESTED = "FIND_HISTORY_REQUESTED";
export const FIND_HISTORY_SUCCEEDED = "FIND_HISTORY_SUCCEEDED";
export const FIND_HISTORY_FAILED = "FIND_HISTORY";

export const REMOVE_CHANGE_REQUESTED = "REMOVE_CHANGE_REQUESTED";
export const REMOVE_CHANGE_SUCCEEDED = "REMOVE_CHANGE_SUCCEEDED";
export const REMOVE_CHANGE_FAILED = "REMOVE_CHANGE_FAILED";

// Users actionTypes
export const LOAD_USERS = "LOAD_USERS";
export const CREATE_USER = "CREATE_USER";

// Account actionTypes
export const GET_ACCOUNT_REQUESTED = "GET_ACCOUNT_REQUESTED";
export const GET_ACCOUNT_SUCCEEDED = "GET_ACCOUNT_SUCCEEDED";
export const GET_ACCOUNT_FAILED = "GET_ACCOUNT_FAILED";

// Administrative setting actionTypes
export const LOAD_SETTINGS = "LOAD_SETTINGS";
export const UPDATE_SETTING = "UPDATE_SETTING";
export const SET_SETTING = "SET_SETTING";



