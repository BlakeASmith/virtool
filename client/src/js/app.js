/**
 * Created by igboyes on 03/03/15.
 */

var React = require('react');
var ReactDOM = require('react-dom');
var Start = require('./components/Start.jsx');

document.write(require('../css/bootstrap.css'));
document.write(require('../css/font.css'));
document.write(require('../css/perfect-scrollbar.css'));
document.write(require('../css/typeahead.css'));
document.write(require('../css/graphics.css'));
document.write(require('../css/style.css'));

ReactDOM.render(React.createElement(Start), document.getElementById('app-container'));