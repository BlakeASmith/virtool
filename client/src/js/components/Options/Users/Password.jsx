/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports AdminChangePassword
 */

'use strict';

var React = require('react');
var Alert = require('react-bootstrap/lib/Alert');
var Panel = require('react-bootstrap/lib/Panel');
var Table = require('react-bootstrap/lib/Table');
var ButtonToolbar = require('react-bootstrap/lib/ButtonToolbar');

var Flex = require('virtool/js/components/Base/Flex.jsx');
var Icon = require('virtool/js/components/Base/Icon.jsx');
var Input = require('virtool/js/components/Base/Input.jsx');
var Checkbox = require('virtool/js/components/Base/Checkbox.jsx');
var PushButton = require('virtool/js/components/Base/PushButton.jsx');
var RelativeTime = require('virtool/js/components/Base/RelativeTime.jsx');

/**
 * The password change form subcomponent of the component exported by the module.
 */
var Change = React.createClass({

    getInitialState: function () {
        return {
            new: '',
            confirm: '',
            failed: false,
            pendingChange: false
        }
    },

    handleChange: function (event) {
        var data = {};
        data[event.target.name] = event.target.value;
        this.setState(data);
    },

    /**
     * Called when the password is submitted. Sends the new password data to the server.
     *
     * @param event - the submit event, used purely to prevent the default submit action
     */
    submit: function (event) {
        event.preventDefault();

        // The new and confirm fields must contain the same password.
        var match = this.state.new === this.state.confirm;

        // The password must have a minimum length (current defaulted to 4).
        var long = this.state.new.length >= 4;

        if (match && long) {
            this.setState({pendingChange: true} , function () {
                dispatcher.db.users.request('set_password', {
                    // The username and new password to assign for it.
                    _id: this.props._id,
                    new_password: this.state.new
                }).success(function () {
                    this.replaceState(this.getInitialState());
                }, this);
            });
        } else {
            this.setState({failed: true});
        }
    },

    render: function () {

        // Check that the passwords pass the following tests.
        var match = this.state.new === this.state.confirm;
        var long = this.state.new.length >= 4;

        var alert;

        // Show and alert with warnings if the password failed the prerequest checks.
        if (!(match && long) && this.state.failed) {
            alert = (
                <Alert bsStyle='danger' className=''>
                    {long ? null: <li>passwords are not at least 4 characters long</li>}
                    {match ? null: <li>passwords don't match</li>}
                </Alert>
            );
        }

        // Boolean to set whether the submit button is disabled or not.
        var submitDisabled = (!(match && long) && this.state.failed) || this.state.pendingChange;

        return (
            <div>
                <p>
                    <em>Last changed </em>
                    <RelativeTime time={this.props.last_password_change} em={true} />
                </p>
                <form onSubmit={this.submit}>
                    <Flex>
                        <Flex.Item grow={1}>
                            <Input
                                type='password'
                                name="new"
                                placeholder='New Password'
                                value={this.state.new}
                                onChange={this.handleChange}
                                disabled={this.state.pendingChange}
                            />
                        </Flex.Item>

                        <Flex.Item grow={1} pad>
                            <Input
                                type='password'
                                name="confirm"
                                placeholder='Confirm Password'
                                value={this.state.confirm}
                                onChange={this.handleChange}
                                disabled={this.state.pendingChange}
                            />
                        </Flex.Item>

                        <Flex.Item pad>
                            <PushButton onClick={this.clear} disabled={this.state.pendingChange}>
                                Clear
                            </PushButton>
                        </Flex.Item>

                        <Flex.Item pad>
                            <PushButton type='submit' bsStyle='primary' disabled={submitDisabled}>
                                <Icon name='floppy' pending={this.state.pendingChange} /> Save
                            </PushButton>
                        </Flex.Item>
                    </Flex>
                </form>
                {alert}
            </div>
        );
    }
});

/**
 * An subcomponent of the password change form that allows the force reset status of the user to be toggled. Consists only of
 * a label checkbox.
 */
var Reset = React.createClass({

    toggle: function () {
        dispatcher.db.users.request('set_force_reset', {
            _id: this.props._id,
            force_reset: !this.props.force_reset
        });
    },

    render: function () {
        return (
            <div className='panel-section'>
                <Checkbox checked={this.props.force_reset} onClick={this.toggle} />
                <span> Force user to reset password on next login.</span>
            </div>
        );
    }

});

/**
 * A parent component to wrap the password change form and reset checkbox with a headed panel.
 */
var AdminChangePassword = React.createClass({

    render: function () {

        return (
            <div>
                <h5><Icon name='lock' /> <strong>Password</strong></h5>

                <Panel>
                    <Change {...this.props} />
                    <Reset {...this.props} />
                </Panel>
            </div>
        );
    }
});

module.exports = AdminChangePassword;