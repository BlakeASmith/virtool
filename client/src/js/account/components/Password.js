import React from "react";
import { connect } from "react-redux";
import { Alert, Col, Panel, Row } from "react-bootstrap";

import { changePassword } from "../actions";
import { Button, Input, RelativeTime } from "../../base";

const getInitialState = () => ({
    oldPassword: "",
    newPassword: "",
    confirmPassword: "",
    errors: []
});

class ChangePassword extends React.Component {

    constructor (props) {
        super(props);
        this.state = getInitialState();
    }

    componentWillReceiveProps (nextProps) {
        if (nextProps.oldPasswordError) {
            this.setState({errors: ["Old password is invalid"]});
        }

        if (nextProps.lastPasswordChange !== this.props.lastPasswordChange) {
            this.setState(getInitialState());
        }
    }

    handleSubmit (e) {
        e.preventDefault();

        const errors = [];

        const minLength = this.props.settings.minimum_password_length;

        if (this.state.confirmPassword.length < minLength || this.state.newPassword.length < minLength) {
            errors.push(`Password must be contain least ${minLength} characters`);
        }

        if (this.state.confirmPassword !== this.state.newPassword) {
            errors.push("New passwords do not match");
        }

        if (errors.length) {
            return this.setState({errors});
        }

        // Set state to show that the user attempted to submit the form.
        this.props.onChangePassword(this.state.oldPassword, this.state.newPassword, this.state.confirmPassword);
    }

    render () {
        if (!this.props.settings) {
            return <div />;
        }

        let errorAlert;

        if (this.state.errors.length) {
            const errorComponents = this.state.errors.map(error =>
                <li key={error}>{error}</li>
            );

            errorAlert = (
                <Alert bsStyle="danger">
                    {errorComponents}
                </Alert>
            );
        }

        return (
            <Panel header="Password">
                <form onSubmit={this.handleSubmit}>
                    <Input
                        label="Old Password"
                        type="password"
                        value={this.state.oldPassword}
                        onChange={(e) => this.setState({oldPassword: e.target.value, errors: []})}
                    />
                    <Input
                        label="New password"
                        type="password"
                        value={this.state.newPassword}
                        onChange={(e) => this.setState({newPassword: e.target.value, errors: []})}
                    />
                    <Input
                        label="Confirm New Password"
                        type="password"
                        value={this.state.confirmPassword}
                        onChange={(e) => this.setState({confirmPassword: e.target.value, errors: []})}
                    />

                    {errorAlert}

                    <div style={{marginTop: "20px"}}>
                        <Row>
                            <Col xs={12} md={6} className="text-muted">
                                Last changed <RelativeTime time={this.props.lastPasswordChange} />
                            </Col>
                            <Col xs={12} md={6}>
                                <Button type="submit" bsStyle="primary" icon="floppy" pullRight>
                                    Change
                                </Button>
                            </Col>
                        </Row>
                    </div>
                </form>
            </Panel>
        );
    }
}

const mapStateToProps = (state) => ({
    lastPasswordChange: state.account.last_password_change,
    oldPasswordError: state.account.oldPasswordError,
    settings: state.settings.data
});

const mapDispatchToProps = (dispatch) => ({

    onChangePassword: (oldPassword, newPassword) => {
        dispatch(changePassword(oldPassword, newPassword));
    }

});

const Container = connect(mapStateToProps, mapDispatchToProps)(ChangePassword);

export default Container;
