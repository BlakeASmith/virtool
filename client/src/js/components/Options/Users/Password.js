/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports AdminChangePassword
 */

import React from "react";
import { Alert, Panel } from "react-bootstrap";
import { Flex, FlexItem, Icon, Input, Checkbox, Button, RelativeTime } from "virtool/js/components/Base";

const getChangeState = () => ({
    password: "",
    confirm: "",
    failed: false,
    pendingChange: false
});

/**
 * The password change form subcomponent of the component exported by the module.
 */
class Change extends React.PureComponent {

    constructor (props) {
        super(props);
        this.state = getChangeState();
    }

    static propTypes = {
        _id: React.PropTypes.string,
        last_password_change: React.PropTypes.string
    };

    handleChange = (event) => {
        let data = {};
        data[event.target.name] = event.target.value;
        this.setState(data);
    };

    isMatch = () => this.state.password === this.state.confirm;
    isLong = () => this.state.password.length >= 4;

    /**
     * Called when the password is submitted. Sends the new password data to the server.
     *
     * @param event - the submit event, used purely to prevent the default submit action
     */
    submit = (event) => {
        event.preventDefault();

        if (this.isMatch() && this.isLong()) {
            this.setState({pendingChange: true} , () => {
                dispatcher.db.users.request("set_password", {_id: this.props._id, new_password: this.state.password})
                    .success(() => {
                        this.setState(getChangeState());
                    });
            });
        } else {
            this.setState({failed: true});
        }
    };

    clear = () => this.setState(getChangeState());

    render () {

        let alert;

        const match = this.isMatch();
        const long = this.isLong();

        // Show and alert with warnings if the password failed the pre-request checks.
        if (!(match && long) && this.state.failed) {
            alert = (
                <Alert bsStyle="danger" className="">
                    {long ? null: <li>passwords are not at least 4 characters long</li>}
                    {match ? null: <li>passwords don"t match</li>}
                </Alert>
            );
        }

        // Boolean to set whether the submit button is disabled or not.
        const submitDisabled = (!(match && long) && this.state.failed) || this.state.pendingChange;

        return (
            <div>
                <p>
                    <em>Last changed </em>
                    <RelativeTime time={this.props.last_password_change} em={true} />
                </p>
                <form onSubmit={this.submit}>
                    <Flex>
                        <FlexItem grow={1}>
                            <Input
                                type="password"
                                name="password"
                                placeholder="New Password"
                                value={this.state.password}
                                onChange={this.handleChange}
                                disabled={this.state.pendingChange}
                            />
                        </FlexItem>

                        <FlexItem grow={1} pad>
                            <Input
                                type="password"
                                name="confirm"
                                placeholder="Confirm Password"
                                value={this.state.confirm}
                                onChange={this.handleChange}
                                disabled={this.state.pendingChange}
                            />
                        </FlexItem>

                        <FlexItem pad>
                            <Button onClick={this.clear} disabled={this.state.pendingChange}>
                                Clear
                            </Button>
                        </FlexItem>

                        <FlexItem pad>
                            <Button type="submit" bsStyle="primary" disabled={submitDisabled}>
                                <Icon name="floppy" pending={this.state.pendingChange} /> Save
                            </Button>
                        </FlexItem>
                    </Flex>
                </form>
                {alert}
            </div>
        );
    }
}

/**
 * An subcomponent of the password change form that allows the force reset status of the user to be toggled. Consists
 * only of a label checkbox.
 */
class Reset extends React.PureComponent {

    static propTypes = {
        _id: React.PropTypes.string,
        forceReset: React.PropTypes.bool
    };

    toggle = () => {
        dispatcher.db.users.request("set_force_reset", {
            _id: this.props._id,
            force_reset: !this.props.forceReset
        });
    };

    render = () => (
        <div className="panel-section">
            <Checkbox checked={this.props.forceReset} onClick={this.toggle} />
            <span> Force user to reset password on next login.</span>
        </div>
    );

}

/**
 * A parent component to wrap the password change form and reset checkbox with a headed panel.
 */
const AdminChangePassword = (props) => (
    <div>
        <h5><Icon name="lock" /> <strong>Password</strong></h5>

        <Panel>
            <Change {...props} />
            <Reset _id={props._id} forceReset={props.force_reset} />
        </Panel>
    </div>
);

AdminChangePassword.propTypes = {
    _id: React.PropTypes.string,
    force_reset: React.PropTypes.bool
};

export default AdminChangePassword;
