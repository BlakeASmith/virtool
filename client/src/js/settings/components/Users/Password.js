/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports AdminChangePassword
 */

import React from "react";
import { connect } from "react-redux";
import { Alert, Panel } from "react-bootstrap";

import { setForceReset, clearSetPassword } from "../../users/actions";
import { FlexItem, Icon, Input, Checkbox, Button, RelativeTime } from "virtool/js/components/Base";

class Password extends React.PureComponent {

    static propTypes = {
        userId: React.PropTypes.string,

        password: React.PropTypes.string,
        confirm: React.PropTypes.string,
        lastPasswordChange: React.PropTypes.string,
        passwordChangePending: React.PropTypes.bool,
        onChangePassword: React.PropTypes.func,
        onChangeConfirm: React.PropTypes.func,

        forceReset: React.PropTypes.bool,
        onSetForceReset: React.PropTypes.func
    };

    isMatch = () => this.state.password === this.state.confirm;
    isLong = () => this.state.password.length >= 4;

    /**
     * Called when the password is submitted. Sends the new password data to the server.
     *
     * @param event - the submit event, used purely to prevent the default submit action
     */
    submit = (event) => {
        event.preventDefault();

        if (this.isMatch() && this.isLong()) {
            this.setState({pendingChange: true} , () => {
                dispatcher.db.users.request("set_password", {_id: this.props._id, new_password: this.state.password})
                    .success(() => {
                        this.setState(getChangeState());
                    });
            });
        } else {
            this.setState({failed: true});
        }
    };

    render () {

        let alert;

        /*

        const match = this.isMatch();
        const long = this.isLong();

        // Show and alert with warnings if the password failed the pre-request checks.
        if (!(match && long) && this.state.failed) {
            alert = (
                <Alert bsStyle="danger" className="">
                    {long ? null: <li>passwords are not at least 4 characters long</li>}
                    {match ? null: <li>passwords don"t match</li>}
                </Alert>
            );
        }

        // Boolean to set whether the submit button is disabled or not.
        const submitDisabled = (!(match && long) && this.state.failed) || this.state.pendingChange;

        */

        const submitDisabled = this.props.passwordChangePending;

        return (
            <div>
                <h5><Icon name="lock" /> <strong>Password</strong></h5>

                <Panel>
                    <p>
                        <em>Last changed </em>
                        <RelativeTime time={this.props.lastPasswordChange} em={true} />
                    </p>

                    <form onSubmit={this.submit}>
                        <div className="toolbar">
                            <FlexItem grow={1}>
                                <Input
                                    type="password"
                                    name="password"
                                    placeholder="New Password"
                                    value={this.props.password}
                                    onChange={this.props.onChangePassword}
                                    disabled={this.props.passwordChangePending}
                                />
                            </FlexItem>

                            <FlexItem grow={1}>
                                <Input
                                    type="password"
                                    name="confirm"
                                    placeholder="Confirm Password"
                                    value={this.props.confirm}
                                    onChange={this.props.onChangeSet}
                                    disabled={this.props.passwordChangePending}
                                />
                            </FlexItem>

                            <Button onClick={this.props.clear} disabled={this.props.passwordChangePending}>
                                Clear
                            </Button>

                            <Button type="submit" bsStyle="primary" disabled={submitDisabled}>
                                <Icon name="floppy" pending={this.props.passwordChangePending} /> Save
                            </Button>
                        </div>
                    </form>

                    {alert}

                    <div className="panel-section">
                        <Checkbox
                            label="Force user to reset password on next login"
                            checked={this.props.forceReset}
                            onClick={() => {this.props.onSetForceReset(this.props.userId, !this.props.forceReset)}}
                        />
                    </div>
                </Panel>
            </div>

        );
    }
}

const mapStateToProps = (state) => {
    return {
        userId: state.users.activeId,

        password: state.users.password,
        confirm: state.users.confirm,
        passwordChangePending: state.users.passwordChangePending,
        lastPasswordChange: state.users.activeData.last_password_change,

        forceReset: state.users.activeData.force_reset
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        onClear: () => {
            dispatch(clearSetPassword())
        },

        onChangeSetPassword: (password, confirm) => {
            dispatch(changeSetPassword(password, confirm));
        },

        onSetForceReset: (userId, enabled) => {
            dispatch(setForceReset(userId, enabled));
        }
    }
};

const Container = connect(mapStateToProps, mapDispatchToProps)(Password);

export default Container;
