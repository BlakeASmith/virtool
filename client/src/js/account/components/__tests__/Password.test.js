import { CHANGE_ACCOUNT_PASSWORD, CLEAR_ERROR } from "../../../app/actionTypes";
import { ChangePassword, mapStateToProps, mapDispatchToProps } from "../Password";

describe("<ChangePassword />", () => {
    let props;

    beforeEach(() => {
        props = {
            error: null,
            lastPasswordChange: "2018-02-14T12:00:00.000000Z",
            minimumLength: 8,
            ready: true,
            onChangePassword: jest.fn(),
            onClearError: jest.fn()
        };
    });

    it("should render", () => {
        const wrapper = shallow(<ChangePassword {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render with error", () => {
        props.error = {
            status: 400,
            message: "existing error"
        };
        const wrapper = shallow(<ChangePassword {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render placeholder when [props.ready=false]", () => {
        props.ready = false;
        const wrapper = shallow(<ChangePassword {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should call onClearError() input changes and there is a pre-existing error", () => {
        props.error = {
            status: 422,
            message: "existing error"
        };
        const wrapper = shallow(<ChangePassword {...props} />);
        const e = {
            target: {
                name: "newPassword",
                value: "foobar"
            }
        };
        wrapper
            .find("InputError")
            .at(1)
            .simulate("change", e);
        expect(props.onClearError).toHaveBeenCalled();
    });

    describe("handleSubmit()", () => {
        let expectedState;

        beforeEach(() => {
            expectedState = {
                oldPassword: "",
                newPassword: "foobar123",
                confirmPassword: "foobar123",
                error: null,
                errorConfirmPassword: "",
                errorNewPassword: "Passwords must contain at least 8 characters",
                errorOldPassword: "Please provide your old password"
            };
        });

        it("should call onSubmit when input values are valid", () => {
            const wrapper = shallow(<ChangePassword {...props} />);
            wrapper.setState({
                oldPassword: "hello1world2",
                newPassword: "foobar123",
                confirmPassword: "foobar123"
            });
            const e = {
                preventDefault: jest.fn()
            };
            wrapper.find("form").simulate("submit", e);
            expect(e.preventDefault).toHaveBeenCalled();
            expect(wrapper.state()).toEqual({
                ...expectedState,
                errorNewPassword: "",
                errorOldPassword: "",
                oldPassword: "hello1world2"
            });
            expect(props.onChangePassword).toHaveBeenCalledWith("hello1world2", "foobar123");
        });

        it("should set error if old password field is empty", () => {
            const wrapper = shallow(<ChangePassword {...props} />);
            wrapper.setState({
                newPassword: "foobar123",
                confirmPassword: "foobar123"
            });
            const e = {
                preventDefault: jest.fn()
            };
            wrapper.find("form").simulate("submit", e);
            expect(e.preventDefault).toHaveBeenCalled();
            expect(wrapper.state()).toEqual({
                ...expectedState,
                errorNewPassword: ""
            });
        });

        it("should set error if new passwords don't match", () => {
            const wrapper = shallow(<ChangePassword {...props} />);
            wrapper.setState({
                oldPassword: "hello1world2",
                newPassword: "foobar124",
                confirmPassword: "foobar123"
            });
            const e = {
                preventDefault: jest.fn()
            };
            wrapper.find("form").simulate("submit", e);
            expect(e.preventDefault).toHaveBeenCalled();
            expect(wrapper.state()).toEqual({
                ...expectedState,
                newPassword: "foobar124",
                oldPassword: "hello1world2",
                errorConfirmPassword: "New passwords do not match",
                errorNewPassword: "",
                errorOldPassword: ""
            });
        });

        it("should set error if new passwords are too short", () => {
            const wrapper = shallow(<ChangePassword {...props} />);
            wrapper.setState({
                oldPassword: "hello1world2",
                newPassword: "foobar",
                confirmPassword: "foobar"
            });
            const e = {
                preventDefault: jest.fn()
            };
            wrapper.find("form").simulate("submit", e);
            expect(e.preventDefault).toHaveBeenCalled();
            expect(wrapper.state()).toEqual({
                ...expectedState,
                newPassword: "foobar",
                confirmPassword: "foobar",
                oldPassword: "hello1world2",
                errorConfirmPassword: "",
                errorNewPassword: "Passwords must contain at least 8 characters",
                errorOldPassword: ""
            });
        });
    });
});

describe("mapStateToProps()", () => {
    const expected = {
        lastPasswordChange: "2018-02-14T12:00:00.000000Z",
        minimumLength: 12,
        error: "Passwords do not match",
        ready: true
    };

    let state;

    beforeEach(() => {
        state = {
            account: {
                last_password_change: "2018-02-14T12:00:00.000000Z"
            },
            settings: {
                data: {
                    minimum_password_length: 12
                }
            },
            errors: {
                CHANGE_ACCOUNT_PASSWORD_ERROR: "Passwords do not match"
            }
        };
    });

    it("should return props with error", () => {
        const props = mapStateToProps(state);
        expect(props).toEqual({ ...expected });
    });

    it("should return props with no error", () => {
        state.errors = {};
        const props = mapStateToProps(state);
        expect(props).toEqual({ ...expected, error: "" });
    });

    it("should return props when settings unavailable", () => {
        state.settings.data = null;
        const props = mapStateToProps(state);
        expect(props).toEqual({ ...expected, ready: false, minimumLength: undefined });
    });
});

describe("mapDispatchToProps()", () => {
    let dispatch;
    let props;

    beforeEach(() => {
        dispatch = jest.fn();
        props = mapDispatchToProps(dispatch);
    });

    it("should return onChangePassword() in props", () => {
        const oldPassword = "foo";
        const newPassword = "bar";
        props.onChangePassword(oldPassword, newPassword);
        expect(dispatch).toHaveBeenCalledWith({
            type: CHANGE_ACCOUNT_PASSWORD.REQUESTED,
            oldPassword,
            newPassword
        });
    });

    it("should return onClearError() in props", () => {
        props.onClearError();
        expect(dispatch).toHaveBeenCalledWith({
            type: CLEAR_ERROR,
            error: "CHANGE_ACCOUNT_PASSWORD_ERROR"
        });
    });
});
