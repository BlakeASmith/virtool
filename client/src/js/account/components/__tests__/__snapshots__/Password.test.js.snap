// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChangePassword /> should render 1`] = `
<Box__BoxGroup>
  <Box__BoxGroupHeader>
    <h2>
      Password
    </h2>
  </Box__BoxGroupHeader>
  <Password__FormButton>
    <form
      onSubmit={[Function]}
    >
      <InputError
        error=""
        label="Old Password"
        name="oldPassword"
        onChange={[Function]}
        type="password"
        value=""
      />
      <InputError
        error=""
        label="New password"
        name="newPassword"
        onChange={[Function]}
        type="password"
        value=""
      />
      <InputError
        error=""
        label="Confirm New Password"
        name="confirmPassword"
        onChange={[Function]}
        type="password"
        value=""
      />
      <Password__ChangePasswordFooter>
        <span>
          Last changed 
          <RelativeTime
            time="2018-02-14T12:00:00.000000Z"
          />
        </span>
        <SaveButton
          altText="Change"
          pullRight={true}
        />
      </Password__ChangePasswordFooter>
    </form>
  </Password__FormButton>
</Box__BoxGroup>
`;

exports[`<ChangePassword /> should render placeholder when [props.ready=false] 1`] = `<LoadingPlaceholder />`;

exports[`<ChangePassword /> should render with error 1`] = `
<Box__BoxGroup>
  <Box__BoxGroupHeader>
    <h2>
      Password
    </h2>
  </Box__BoxGroupHeader>
  <Password__FormButton>
    <form
      onSubmit={[Function]}
    >
      <InputError
        error="existing error"
        label="Old Password"
        name="oldPassword"
        onChange={[Function]}
        type="password"
        value=""
      />
      <InputError
        error=""
        label="New password"
        name="newPassword"
        onChange={[Function]}
        type="password"
        value=""
      />
      <InputError
        error=""
        label="Confirm New Password"
        name="confirmPassword"
        onChange={[Function]}
        type="password"
        value=""
      />
      <Password__ChangePasswordFooter>
        <span>
          Last changed 
          <RelativeTime
            time="2018-02-14T12:00:00.000000Z"
          />
        </span>
        <SaveButton
          altText="Change"
          pullRight={true}
        />
      </Password__ChangePasswordFooter>
    </form>
  </Password__FormButton>
</Box__BoxGroup>
`;
